cross-version-test/cross_version_test/cli.py

```diff
@@ -0,0 +1,276 @@
+        install_script.write_text(SHELL_SCRIPT_TEMPTLATE.render(cmd=job.install))
+        run_script.write_text(SHELL_SCRIPT_TEMPTLATE.render(cmd=job.run))
+        entrypoint_script.write_text(entrypoint)
+        # change permission
+        os.chmod(install_script, 0o777)
+        os.chmod(run_script, 0o777)
+        os.chmod(entrypoint_script, 0o777)
+        # Create dockerfile
+        content = DOCKERFILE_TEMPLATE.render(
+            base_image_name=c.DOCKER_BASE_IMAGE,
+            mlflow_requirements_file=c.MLFLOW_REQUIREMENTS_FILE,
+            install_script=c.INSTALL_SCRIPT,
+            run_script=c.RUN_SCRIPT,
+            entrypoint_script=c.ENTRYPOINT_SCRIPT,
+            python_version="3.7",
+        )
+        job_dir.joinpath(c.DOCKERFILE).write_text(content)
+        copy_tree("requirements", str(job_dir.joinpath("requirements")))
+
+    docker_compose_content = DOCKER_COMPOSE_TEMPLATE.render(jobs=[j.name for j in sorted_jobs])
+    DOCKER_COMPOSE_FILE.write_text(docker_compose_content)
+    click.echo("Done!")
+
+
+def list_jobs() -> t.List[str]:
+    docker_compose = DockerCompose(DOCKER_COMPOSE_FILE)
+    return docker_compose.config("--services", capture_output=True, check=True).stdout.splitlines()
+
+
+@cli.command(name="list")
+def list_command():
+    click.echo("\n".join(list_jobs()))
+
+
+@cli.command(context_settings=dict(ignore_unknown_options=True))
+@click.option(
+    "--pattern",
+    "-p",
+    required=False,
+    help="If specified, only run jobs that matches the given pattern (regular expression). "
+    "For example, 'xgboost_.+_autologging' matches xgboost autologging jobs.",
+)
+@click.option(
+    "--no-cache", is_flag=True, default=False, help="Do not use cache when building images"
+)
+def build(pattern: t.Optional[str], no_cache: bool):
+    jobs = list_jobs()
+
+    no_cache_opt = ["--no-cache"] if no_cache else []
+
+    # Build base image
+    reporter = Reporter()
+    reporter.write_separator("Building base image")
+    this_file_dir = Path(__file__).parent
+    base_dockerfile = this_file_dir.joinpath("base_image", "Dockerfile")
+    run_cmd(
+        [
+            "docker",
+            "build",
+            "-f",
+            base_dockerfile,
+            *no_cache_opt,
+            "-t",
+            c.DOCKER_BASE_IMAGE,
+            this_file_dir,
+        ],
+        check=True,
+    )
+
+    if pattern:
+        jobs = [j for j in jobs if re.compile(pattern).search(j)]
+
+    docker_compose = DockerCompose(DOCKER_COMPOSE_FILE)
+    for job in jobs:
+        reporter.write_separator(f"Building {job}")
+        prc = docker_compose.build(*no_cache_opt, job, check=False)
+        reporter.add_result(Result(job, prc.success))
+
+    reporter.write_separator("Results")
+    reporter.report_results()
+    failed = not reporter.all_success()
+    exit(failed)
+
+
+@cli.command(context_settings=dict(ignore_unknown_options=True))
+@click.option(
```

Comment:
Did you want to put the repeated option args into a utility module or is that overkill for just a single duplicated statement?

Reply:
I think it's overkill. I'll just put the `--pattern` option in a variable.